File: Ayeon/[1월 17일] 정렬/2751.py
import sys

def merge_sort(unsorted):
    if len(unsorted) <= 1:
        return unsorted
    
    mid=len(unsorted)//2
    left=unsorted[:mid]
    right=unsorted[mid:]

    left1=merge_sort(left)
    right1=merge_sort(right)
    return merge(left1, right1)

def merge(left, right):
    i=0
    j=0
    sorted=[]

    while((i<len(left)) & (j<len(right))):
        if left[i] < right[j]:
            sorted.append(left[i])
            i+=1
        else:
            sorted.append(right[j])
            j+=1
    
    while(i<len(left)):
        sorted.append(left[i])
        i+=1

    while(j<len(right)):
        sorted.append(right[j])
        j+=1
    
    return sorted

N=int(input())
a=[int(sys.stdin.readline().rstrip()) for i in range(N)]

#a.sort()
a=merge_sort(a)

for i in a:
    sys.stdout.write(str(i)+'\n')


        

File: Siyeon/2751.cpp
//이유는 모르나 vector이나 iostream 쓰려고 하면 시간초과남..
#include <cstdio>
constexpr auto MAX = 1'000'000;

void merge(int* v, int left, int mid, int right) {
    int a[MAX];
    int i = left, j = mid + 1, k = left, l=0;
    while (i <= mid && j <= right) {
        if (v[i] <= v[j]) {
            a[k++] = v[i++];
        }
        else {
            a[k++] = v[j++];
        }
    }
    if (i > mid) {
        for (int l = j; l <= right; l++) {
            a[k++] = v[l];
        }
    }
    else {
        for (int l = i; l <= mid; l++) {
            a[k++] = v[l];
        }
    }
    for (int l = left; l <= right; l++) {
        v[l] = a[l];
    }
}

void merge_sort(int* v, int left, int right) {
    int mid;
    if (left < right) {
        mid = (left + right) / 2;
        merge_sort(v, left, mid);
        merge_sort(v, mid + 1, right);
        merge(v, left, mid, right);
    }
}

int main() {
    int cnt;
    scanf("%d", &cnt);
    int num[MAX];
    for (int i = 0; i < cnt; i++) {
        scanf("%d", &num[i]);
    }
    merge_sort(num, 0, cnt - 1);
    for (int i = 0; i < cnt; i++) {
        printf("%d\n", num[i]);
    }

}
